### DISCUSSION TITLE
k3s node uses the wrong IP

### DISCUSSION THREAD

**Discussion by sporqist** [k3s-io/k3s#6029] [General]
**Environmental Info:**
K3s Version: 
k3s version v1.24.3+k3s1 (990ba0e8)
go version go1.18.1

Cluster Configuration:
1 Server with embedded etcd, 3 Agents
Nodes are connected via a wireguard mesh.
Server is the only one with a public IP.

**Describe the bug:**
After a cluster-rebuild, i struggle to get my cluster back online. K3s seems to partly ignore the IP assignments via arguments. I start the server node with these args via a systemd unit to make sure k3s only talks via wireguard.
```
ExecStart=/usr/local/bin/k3s \
    server \
        --data-dir /var/lib/rancher/k3s \
        --flannel-iface wg0 \
        --bind-address 10.222.0.3 \
        --node-ip 10.222.0.3 \
        --node-external-ip 10.222.0.3 \
        --advertise-address 10.222.0.3
```

Now, if I check the annotations of the server node, many don't match my arguments:
```
Annotations:        alpha.kubernetes.io/provided-node-ip: 10.222.0.3
                    csi.volume.kubernetes.io/nodeid: {"driver.longhorn.io":"debian-4gb-fsn1-1"}
                    etcd.k3s.cattle.io/node-address: 142.132.135.125
                    etcd.k3s.cattle.io/node-name: debian-4gb-fsn1-1-70e3a4b1
                    flannel.alpha.coreos.com/backend-type: vxlan
                    flannel.alpha.coreos.com/kube-subnet-manager: true
                    flannel.alpha.coreos.com/public-ip: 142.132.135.125
                    k3s.io/external-ip: 10.222.0.3
                    k3s.io/hostname: debian-4gb-fsn1-1
                    k3s.io/internal-ip: 142.132.135.125
                    k3s.io/node-args: ["server","\\"]
...
```

The connected agent nodes all want to connect to this proxy IP: 142.132.135.125 which of course they can't because they can't reach the proxy there and i want them to talk to 10.222.0.3 via wireguard instead and not some public IP anyways. The coredns configmap also lists the wrong IP for the server node.

**Steps To Reproduce:**
- have a k3s cluster
- change your servers IP
- do a cluster-rebuild
- restart k3s
- ?!

**Expected behavior:**
k3s respects the arguments.

**Actual behavior:**
k3s uses some random available IP to ruin my day

**Additional context / logs:**
I tried to do the cluster rebuild with all the args from my systemd unit and the rebuild succeeds but if i do it like this, etcd fails with this error:
```
Failed to test data store connection: this server is a not a member of the etcd cluster. Found [debian-4gb-fsn1-1-6b95212b=https://10.222.0.3:2380], expect: debian-4gb-fsn1-1-6b95212b=https://142.132.135.125:2380
```

**Comment by brandond** [comment]
Why are you setting `--node-ip` and `--node-external-ip` to the same thing? The external IP is meant to be used when you have a public IP address that is NATed to the node's private IP - think EC2 EIPs.

Also, node IPs should not be changed after the fact, as they are intended to be static for the life of the node in the cluster. If you need to change them, you should use `--cluster-reset` to reset the cluster membership to just the current node with the current IP. From there you can re-add additional servers to the cluster with whatever IP address you want.

**Comment by sporqist** [comment]
Thx for clearing up `--node-ip` and `--node-external-ip` for me. As you can probably tell, I still have much to learn about k3s. The docs didn't make it clear for me so i set everything to the wireguard internal IP for testing. Now it looks like this: 
```
ExecStart=/usr/local/bin/k3s \
    server \
        --data-dir /var/lib/rancher/k3s \
        --flannel-iface wg0 \
        --bind-address 10.222.0.3 \
        --node-ip 10.222.0.3 \
        --node-external-ip 142.132.135.125 \
        --advertise-address 10.222.0.3
```

I already did a `--cluster-reset` (called it cluster-rebuild before. sorry about that) but agent nodes still try to use the main nodes public IP instead of the internal one.

From an agent nodes `journalctl -xeu k3s`:
```
level=error msg="Failed to connect to proxy. Empty dialer response" error="dial tcp 142.132.135.125:6443: connect: connection timed out"
level=error msg="Remotedialer proxy error" error="dial tcp 142.132.135.125:6443: connect: connection timed out"
level=info msg="Connecting to proxy" url="wss://142.132.135.125:6443/v1-k3s/connect"
``` 

Coredns's configmap also lists the wrong IP:
```yaml
NodeHosts: |
  142.132.135.125 debian-4gb-fsn1-1
  10.222.1.1 kube-1
  10.222.1.101 kube-101
  10.222.0.1 sporq
```

**Comment by brandond** [comment]
> agent nodes still try to use the main nodes public IP instead of the internal one.

Yes, that is expected. If you set an external address, it is expected that this address can be reached by all the agents.

Rather than try to get all the nodes running over an existing wireguard implementation, have you tried just using the wireguard-native backend and letting flannel build the wireguard mesh for you?

**Comment by sporqist** [comment]
I will try using flannels wireguard backend tomorrow. Right now the only open ports on the main server, reachable from the internet, are HTTP, HTTPS and the wireguard port. That's why the nodes can only reach the apiserver via the wireguard network. The server address I give the agent nodes to connect to is also the internal wireguard address `10.222.0.3`.

Check this output from `kubectl get node debian-4gb-fsn1-1 -o yaml` (the main node with embedded etcd):
```yaml
status:
  addresses:
  - address: 10.222.0.3
    type: ExternalIP
  - address: 142.132.135.125
    type: InternalIP
```
The flannel backend address from wg0 is used as external IP and the public IP is being used as internal IP.

Is this why the nodes don't use the wireguard IP? Just found this in the logs of an agent:
```
level=info msg="Found ip 10.222.1.1 from iface wg0"
level=info msg="Starting k3s agent v1.24.3+k3s1 (990ba0e8)"
level=info msg="Running load balancer k3s-agent-load-balancer 127.0.0.1:6444 -> [142.132.135.125:6443 10.222.0.3:6443]"
...
level=info msg="Containerd is now running"
level=info msg="Getting list of apiserver endpoints from server"
level=info msg="Updating load balancer k3s-agent-load-balancer default server address -> 142.132.135.125:6443"
level=info msg="Connecting to proxy" url="wss://142.132.135.125:6443/v1-k3s/connect"
```
First, the internal IP is listed as load-balancer IP, then it gets updated to only the public IP.

I know, it was working with k3s v1.22.7+k3s1.

**Reply by brandond** [reply]
Yes, that's correct. It initially connects to the address you specified via the `--server` URI. Once it is joined to the cluster, the load-balancer is reconfigured to directly connect to the servers via their public IP addresses. If you're configuring an external IP, it is expected that the server can be reached at that address.

**Reply by sporqist** [reply]
So how do i set things right and teach k3s what its internal and external IPs are? They are still swapped. This is a problem because of multiple reasons. For one, Ingresses now listen to the internal IP and are not reachable from the outside where the ingress should serve traffic.

**Reply by brandond** [reply]
Are you still telling it to advertise the internal IP instead of the external?

**Reply by brandond** [reply]
Also, changing nodes IPs after the fact isn't well supported by Kubernetes. They are only properly set once, when the node joins the cluster. If you're playing with the flags to try to get what you want, I would be sure to uninstall/reinstall between attempts.

**Reply by sporqist** [reply]
I deleted the cluster and initialized a few new ones (uninstalling between the attempts) to find out how to make k3s use the wireguard IP for the API. Nothing, including the `--bind-address` option, worked out for me. [This](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/) resource helped me a little bit with understanding `--advertise-address` and `--bind-address` with a little more elaborate descriptions than in k3s's docs.

I scrapped the idea to make k3s use my own wireguard network, as you suggested. Now, the cluster is running fine, using flannels own wireguard backend. Had to open more Ports on the Firewall than with the old setup but :man_shrugging:

Anyways, thanks for your help. I didn't find a way to buy you a coffee but here is a virtual one: :coffee:

**Reply by sporqist** [reply]
Sooo flannels wireguard backend didn't end up working. Some things like cert-managers tried to call webhooks around the wireguard tunnel and that of course didn't work because these ports were closed by a firewall.. I checked my broken config from before and realised, there was a space behind one of my line escapes in the config which resulted in k3s not receiving the arguments I gave it. :man_facepalming: 

```
ExecStart=/usr/local/bin/k3s \
    server \ <- The space character that cost me hours of despair
        --cluster-init \
        --data-dir={{ k3s_server_location }} \
        --node-external-ip {{ ansible_default_ipv4['address'] }} \
        --node-ip {{ ansible_wg0.ipv4.address }} \
        --flannel-iface wg0
```

PS: I tried it with an identical (wireguard) IP for `--node-ip` and `--node-external-ip` and with the correct ones (wireguard and public IP) and both setups worked fine, except I wasn't able to reach the testing website behind my ingress on the public IP, when I set the wireguard IP for both options.
